### PR TITLE
core: add interrupt framework

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Adds interrupt frameworks and adjusts gic driver to fit in.

Update plat-vexpress and sunxi platforms to initialize gic with slightly
modified interface.

Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (QEMU, FVP)
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
The interrupt framework doesn't currently have any locks. That's not a problem for the moment, but if we add handlers after boot time it will become a problem. Should we address this now or wait until the system is more dynamic?

File: core/drivers/gic.c [hunk 1]
```diff
@@ -152,154 +162,205 @@ void gic_init(vaddr_t gicc_base, vaddr_t gicd_base)
...
+	gic_write_eoir(gd, iar);
+}
+
+static void gic_op_add(struct itr_chip *chip, size_t it,
+		       uint32_t flags __unused)
+{
+	struct gic_data *gd = container_of(chip, struct gic_data, chip);
+
+	gic_it_add(gd, it);
+	gic_it_set_cpu_mask(gd, it, 0xff);
```
Comment (pascal-brand-st-dev) [inline core/drivers/gic.c]: `0xff` used to be `gic_it_get_target(0)` is it an issue to have this hard-coded value?

File: core/arch/arm/plat-vexpress/main.c [hunk 2]
```diff
@@ -66,58 +69,32 @@ static const struct thread_handlers handlers = {
...
-	gic_it_set_cpu_mask(IT_CONSOLE_UART, gic_it_get_target(0));
-	gic_it_set_prio(IT_CONSOLE_UART, 0x1);
-	gic_it_enable(IT_CONSOLE_UART);
-#endif
-#endif
-}
-#elif PLATFORM_FLAVOR_IS(qemu_virt)
 void main_init_gic(void)
 {
+#if PLATFORM_FLAVOR_IS(fvp) || PLATFORM_FLAVOR_IS(juno) || \
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/plat-vexpress/main.c]: Can it use CFG_ARM32 instead of platform flavor? It can be part of another PR as not directly related to this one.

**Comment by pascal-brand-st-dev** [comment]
About the lock that is missing, I am ok with it now. However, can you add a "Fixme" comment just to remember.

Also, few travis checkpatch errors. Not sure it make sense to fix.

With or without my comments:
`Reviewed-by: Pascal Brand <pascal.brand@linaro.org>`

File: core/drivers/gic.c (hunk 1 above)
Comment (jenswi-linaro) [inline core/drivers/gic.c]: 0xff means any cpu online, while the old evaluated to this cpu only. I'll add a comment.

File: core/arch/arm/plat-vexpress/main.c (hunk 2 above)
Comment (jenswi-linaro) [inline core/arch/arm/plat-vexpress/main.c]: FVP and Juno supports both arm32 and arm64.

**Comment by jenswi-linaro** [comment]
Both checkpatch warnings are false positives in my opinion.

I'll add a warning about locking.

File: core/arch/arm/plat-vexpress/main.c (hunk 2 above)
Comment (pascal-brand-st-dev) [inline core/arch/arm/plat-vexpress/main.c]: Right. Then I would prefer to have a new CFG_WITH_ARM_TF instead of this. But this is only an idea to keep in mind for another PR in case it sounds good.

**Comment by pascal-brand-st-dev** [comment]
`Reviewed-by: Pascal Brand <pascal.brand@linaro.org>`

**Comment by jenswi-linaro** [comment]
Rebased and tag applied.
I've added the comments discussed above. I've also made a small change to plat-sunxi to pass the pointer to `gic_data` in for `r0` for `gic_cpu_init`

**Comment by pascal-brand-st-dev** [comment]
`Reviewed-by: Pascal Brand <pascal.brand@linaro.org>`

**Comment by jenswi-linaro** [comment]
Squashed.
